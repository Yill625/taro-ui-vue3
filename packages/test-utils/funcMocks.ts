import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks