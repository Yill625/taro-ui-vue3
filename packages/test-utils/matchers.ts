import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers