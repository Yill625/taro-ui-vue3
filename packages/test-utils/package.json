{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}