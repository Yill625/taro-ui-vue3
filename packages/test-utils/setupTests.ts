import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()