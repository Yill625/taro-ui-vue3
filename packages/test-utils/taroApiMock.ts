import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
