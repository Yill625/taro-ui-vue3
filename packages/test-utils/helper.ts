import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}