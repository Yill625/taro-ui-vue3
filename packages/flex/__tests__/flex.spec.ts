import { mountFactory } from '@taro-ui-vue3/test-utils/helper'
import { h } from '@vue/runtime-core'
import AtFlex from "../index"
import AtFlexItem from "../item"

describe.each([
  ['wrap', ["wrap", "no-wrap", "wrap-reverse"]],
  ['align', ["start", "end", "center", "stretch", "baseline"]],
  ['justify', ["start", "end", "center", "between", "around"]],
  ['direction', ["row", "column", "row-reverse", "column-reverse"]],
  ['alignContent', ["start", "end", "center", "stretch", "between", "around"]]
])('AtFlex', (propName, propOptions) => {
  it.each(propOptions)(`should render prop ${propName} -- %s`, (propOption) => {
    const wrapper = mountFactory(
      AtFlex,
      undefined,
      { [`${propName}`]: propOption }
    )

    expect(wrapper.element).toMatchSnapshot()
    switch (propName) {
      case 'wrap':
        expect(wrapper.find(`.at-row--${propOption}`).exists()).toBeTruthy()
        break
      case 'alignContent':
        expect(wrapper.find(`.at-row__align-content--${propOption}`).exists()).toBeTruthy()
        break
      default:
        expect(wrapper.find(`.at-row__${propName}--${propOption}`).exists()).toBeTruthy()
        break
    }
  })

  it('should render default slot content', () => {
    const wrapper = mountFactory(
      AtFlex,
      undefined,
      undefined,
      { default: [h('view')] }
    )

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find('.at-row > view').exists()).toBeTruthy()

  })
})

describe('AtFlexItem', () => {
  it.each([
    'isAuto',
    'isWrap'
  ])('should render prop -- %s', (propName) => {
    const wrapper = mountFactory(
      AtFlexItem,
      undefined,
      { [`${propName}`]: true }
    )

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(`.at-col--${propName.toLowerCase().replace('is', '')}`).exists()).toBeTruthy()
  })

  it.each([
    "center", "top", "bottom"
  ])('should render prop align -- %s', (propOption) => {
    const wrapper = mountFactory(
      AtFlexItem,
      undefined,
      { align: propOption }
    )

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(`.at-col__align--${propOption}`).exists()).toBeTruthy()
  })

  it.each([
    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12
  ])('should render prop offset -- %s', (propOption) => {
    const wrapper = mountFactory(
      AtFlexItem,
      undefined,
      { offset: propOption }
    )

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(`.at-col__offset-${propOption}`).exists()).toBeTruthy()
  })

  it.each([
    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12
  ])('should render prop size -- %s', (propOption) => {
    const wrapper = mountFactory(
      AtFlexItem,
      undefined,
      { size: propOption }
    )

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find(`.at-col-${propOption}`).exists()).toBeTruthy()
  })

  it('should render default slot content', () => {
    const wrapper = mountFactory(
      AtFlexItem,
      undefined,
      undefined,
      { default: [h('view')] }
    )

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find('.at-col > view').exists()).toBeTruthy()

  })
})