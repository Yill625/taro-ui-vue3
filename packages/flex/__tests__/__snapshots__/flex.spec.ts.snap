// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtFlex should render default slot content 1`] = `
<view
  class="at-row"
>
  <view />
</view>
`;

exports[`AtFlex should render default slot content 2`] = `
<view
  class="at-row"
>
  <view />
</view>
`;

exports[`AtFlex should render default slot content 3`] = `
<view
  class="at-row"
>
  <view />
</view>
`;

exports[`AtFlex should render default slot content 4`] = `
<view
  class="at-row"
>
  <view />
</view>
`;

exports[`AtFlex should render default slot content 5`] = `
<view
  class="at-row"
>
  <view />
</view>
`;

exports[`AtFlex should render prop align -- baseline 1`] = `
<view
  class="at-row at-row__align--baseline"
/>
`;

exports[`AtFlex should render prop align -- center 1`] = `
<view
  class="at-row at-row__align--center"
/>
`;

exports[`AtFlex should render prop align -- end 1`] = `
<view
  class="at-row at-row__align--end"
/>
`;

exports[`AtFlex should render prop align -- start 1`] = `
<view
  class="at-row at-row__align--start"
/>
`;

exports[`AtFlex should render prop align -- stretch 1`] = `
<view
  class="at-row at-row__align--stretch"
/>
`;

exports[`AtFlex should render prop alignContent -- around 1`] = `
<view
  class="at-row at-row__align-content--around"
/>
`;

exports[`AtFlex should render prop alignContent -- between 1`] = `
<view
  class="at-row at-row__align-content--between"
/>
`;

exports[`AtFlex should render prop alignContent -- center 1`] = `
<view
  class="at-row at-row__align-content--center"
/>
`;

exports[`AtFlex should render prop alignContent -- end 1`] = `
<view
  class="at-row at-row__align-content--end"
/>
`;

exports[`AtFlex should render prop alignContent -- start 1`] = `
<view
  class="at-row at-row__align-content--start"
/>
`;

exports[`AtFlex should render prop alignContent -- stretch 1`] = `
<view
  class="at-row at-row__align-content--stretch"
/>
`;

exports[`AtFlex should render prop direction -- column 1`] = `
<view
  class="at-row at-row__direction--column"
/>
`;

exports[`AtFlex should render prop direction -- column-reverse 1`] = `
<view
  class="at-row at-row__direction--column-reverse"
/>
`;

exports[`AtFlex should render prop direction -- row 1`] = `
<view
  class="at-row at-row__direction--row"
/>
`;

exports[`AtFlex should render prop direction -- row-reverse 1`] = `
<view
  class="at-row at-row__direction--row-reverse"
/>
`;

exports[`AtFlex should render prop justify -- around 1`] = `
<view
  class="at-row at-row__justify--around"
/>
`;

exports[`AtFlex should render prop justify -- between 1`] = `
<view
  class="at-row at-row__justify--between"
/>
`;

exports[`AtFlex should render prop justify -- center 1`] = `
<view
  class="at-row at-row__justify--center"
/>
`;

exports[`AtFlex should render prop justify -- end 1`] = `
<view
  class="at-row at-row__justify--end"
/>
`;

exports[`AtFlex should render prop justify -- start 1`] = `
<view
  class="at-row at-row__justify--start"
/>
`;

exports[`AtFlex should render prop wrap -- no-wrap 1`] = `
<view
  class="at-row at-row--no-wrap"
/>
`;

exports[`AtFlex should render prop wrap -- wrap 1`] = `
<view
  class="at-row at-row--wrap"
/>
`;

exports[`AtFlex should render prop wrap -- wrap-reverse 1`] = `
<view
  class="at-row at-row--wrap-reverse"
/>
`;

exports[`AtFlexItem should render default slot content 1`] = `
<view
  class="at-col"
>
  <view />
</view>
`;

exports[`AtFlexItem should render prop -- isAuto 1`] = `
<view
  class="at-col--auto at-col"
/>
`;

exports[`AtFlexItem should render prop -- isWrap 1`] = `
<view
  class="at-col--wrap at-col"
/>
`;

exports[`AtFlexItem should render prop align -- bottom 1`] = `
<view
  class="at-col__align--bottom at-col"
/>
`;

exports[`AtFlexItem should render prop align -- center 1`] = `
<view
  class="at-col__align--center at-col"
/>
`;

exports[`AtFlexItem should render prop align -- top 1`] = `
<view
  class="at-col__align--top at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 1 1`] = `
<view
  class="at-col__offset-1 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 2 1`] = `
<view
  class="at-col__offset-2 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 3 1`] = `
<view
  class="at-col__offset-3 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 4 1`] = `
<view
  class="at-col__offset-4 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 5 1`] = `
<view
  class="at-col__offset-5 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 6 1`] = `
<view
  class="at-col__offset-6 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 7 1`] = `
<view
  class="at-col__offset-7 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 8 1`] = `
<view
  class="at-col__offset-8 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 9 1`] = `
<view
  class="at-col__offset-9 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 10 1`] = `
<view
  class="at-col__offset-10 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 11 1`] = `
<view
  class="at-col__offset-11 at-col"
/>
`;

exports[`AtFlexItem should render prop offset -- 12 1`] = `
<view
  class="at-col__offset-12 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 1 1`] = `
<view
  class="at-col-1 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 2 1`] = `
<view
  class="at-col-2 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 3 1`] = `
<view
  class="at-col-3 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 4 1`] = `
<view
  class="at-col-4 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 5 1`] = `
<view
  class="at-col-5 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 6 1`] = `
<view
  class="at-col-6 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 7 1`] = `
<view
  class="at-col-7 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 8 1`] = `
<view
  class="at-col-8 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 9 1`] = `
<view
  class="at-col-9 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 10 1`] = `
<view
  class="at-col-10 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 11 1`] = `
<view
  class="at-col-11 at-col"
/>
`;

exports[`AtFlexItem should render prop size -- 12 1`] = `
<view
  class="at-col-12 at-col"
/>
`;
